### ISSUE TITLE
multiple wayland windows stuck at OpenGL rendering when occluded

### ISSUE THREAD

**Issue by christianrauch** [libsdl-org/SDL#4335]
When multiple Wayland windows are used with OpenGL and one of these windows is entirely occluded by another window, then all windows will not receive any further callbacks at will get stuck in their OpenGL rendering loop.

Reproduce:
1. run `testgl2` with 2 windows: `SDL_VIDEODRIVER=wayland ./testgl2 --windows 2`
2. hide both windows behind another window, e.g. by maximising your browser over both windows
3. select one of the window to appear in the foreground

The foreground window will now show a cube that is stuck at one point in the spinning animation. Only when the other window is at least partially visible, the cubes will rotate again. This also has implications for client-side decoration as it will prevent interaction (resize, close) with that window. Both windows will also continue their animation if they are visible on a thumbnail overview, e.g. in GNOME Shell when the Activity overview is activated by the top-left hot corner or the Meta / Windows key.

The reason for this is that Wayland compositors will inhibit frame callbacks for window that are not visible, so do not update the window when its content is not shown. When multiple windows are part of a render loop, these frame callbacks have to be handled manually.

See:
- https://emersion.fr/blog/2018/wayland-rendering-loop/
- https://github.com/glfw/glfw/issues/1350

**Comment by flibitijibibo** [comment]
I don't think we can mess with the loop at all, so we may have to just set swap interval to 0 (or leave it alone if it's-1, should EGL_EXT_swap_control_tear get approved) so it doesn't block on eglSwapBuffers. I bet this is why I was getting blocked on #4319 for certain compositors... I'll have to try this in some FNA tests.

**Comment by icculus** [comment]
This annoys me because this is going to be sold as Much More Efficient but in reality will probably offer Much More Latency, and certainly Much More Inconvenience.  :/

More to the point, to support this as expected in every game ever written, we'll have to emulate vsync. I guess we should set the real swap interval to 0 unconditionally, and just SDL_Delay() to keep the framerate to 60Hz or something. But it's not going to actually _match_ the hardware vsync in this case.

This is deeply disappointing. I wish this would change in Wayland.

**Comment by flibitijibibo** [comment]
> More to the point, to support this as expected in every game ever written, we'll have to emulate vsync. I guess we should set the real swap interval to 0 unconditionally, and just SDL_Delay() to keep the framerate to 60Hz or something. But it's not going to actually _match_ the hardware vsync in this case.

This may not be necessary - most games have their own fixed step anyway and we know Wayland is going to force tear-free anyway, so perhaps we should just overload SDL_EGL_SetSwapInterval to force 0 and return -1 for lack of interval support, and the game will handle the rest.

> This is deeply disappointing. I wish this would change in Wayland.

Agreed... between this and the xdg-surface deadlock this kind of went from 100 to 0 real quick.

**Comment by flibitijibibo** [comment]
Update: Turns out my xdg-surface deadlock was just this the whole time. So yeah, we _really_ need to hack eglSwapInterval if we want to keep this from breaking everybody's stuff (FNA's included). As I mentioned earlier the behavior is to never tear even for interval 0, so my vote is to override SDL_EGL_SwapInterval to clamp to 0 with an SDL_SetError explaining the Wayland flaw. @icculus, if this is okay with you I'll write it up ASAP.

**Comment by fooishbar** [comment]
Yes it is annoying - sorry - but it's not quite as bad as it's made out here. You don't need to burn CPU and you don't need to set arbitrary 16ms timeouts which will definitely be out of phase and probably out of time too. There's nothing special and magic about the way that `eglSwapBuffers` blocks which clients can't do: it just calls `wl_surface_frame()` for the compositor to signal when is a good time to submit the next frame. So you can do that too with a swap interval of 0.

There's nothing blocking about it either - that call returns a `wl_callback` whose `done` event will signal when ready - it's just that EGL chooses to block to implement the 'dumbest possible loop of `while (true) { render(); eglSwapBuffers(); }` should throttle and not burn your CPU' part of the EGL API, and many compositors choose to not send frame events to occluded clients because why are you still burning resources generating frames which will never be seen.

So anyway, there you are, just submit the frame request before you call SwapBuffers, use its 'done' event to progress on to the next frame (this is what Mesa does internally during SwapBuffers, but it has previously been during the first draw call of the frame), and if it's not progressing within a given time (the compositor doesn't want you submitting new frames), then since you control the event loop you can decide to just do it anyway after a timeout, or do other things without being blocked, or whatever.

**Comment by christianrauch** [comment]
@icculus @flibitijibibo I don't see a big issue here. I think for multiple parallel render loops, you have to either use multiple `wl_event_queue` to handle the different window states or you let the compositor decide when to draw via the `wl_surface_frame` callbacks as described in https://emersion.fr/blog/2018/wayland-rendering-loop/ by @emersion. There is no need to hack around this issue.

**Comment by icculus** [comment]
Okay, this sounds like there is a way forward; I probably overreacted, so I apologize for that.

@flibitijibibo, it sounds like we can do this without bad hacks after all. Let me know if you want me to do it, otherwise I assume you'll get it.

**Comment by flibitijibibo** [comment]
This still sounds like a hack to me - the implication is that instead of blocking inside eglSwapBuffers we block inside SDL instead? Someone else should cover this, I don't think I understand this idea well enough to write it.

**Comment by icculus** [comment]
For Vulkan, is this a FIFO thing? Does Mailbox work as expected?

**Comment by flibitijibibo** [comment]
Mailbox is fine, it only affects FIFO.

**Comment by fooishbar** [comment]
> This still sounds like a hack to me - the implication is that instead of blocking inside eglSwapBuffers we block inside SDL instead?

That's right. Set your swap interval to 0, call `wl_surface_frame()` before calling `eglSwapBuffers()`, then add an event listener on the `wl_callback` proxy it returns to fire with its `done` event. It will then trigger when the compositor wants you to send a new frame, without blocking anywhere.

The meaning of the frame event is 'the compositor thinks it is a good time for you to start working on your next frame'; compositors like Mutter (GNOME Shell) will not send this completion event whilst the window is fully occluded, since they do not want you to send useless updates which will be discarded, hence there is no 'good time' for this. Xorg sends updates at 1Hz so you do keep vaguely ticking, but that's not super helpful if you've got network traffic or audio processing blocked on EGL returning.

Vulkan doesn't have this issue as `vkAcquireNextImageKHR` takes a timeout, unlike EGL where the API gives you no option but to block indefinitely.

**Comment by flibitijibibo** [comment]
>Vulkan doesn't have this issue as vkAcquireNextImageKHR takes a timeout, unlike EGL where the API gives you no option but to block indefinitely.

I can confirm on Plasma that this does not work as intended. I have updated the test program accordingly:

http://www.flibitijibibo.com/wltest.tar.bz2

To test, simply press the space bar and wait 3 seconds for the window to reappear. This uses the following branch of SDL...

https://github.com/flibitijibibo/SDL/tree/2.0.14-wayland

... and the following branch of FNA3D:

https://github.com/FNA-XNA/FNA3D/

The current FNA3D revision uses UINT64_MAX, but can be modified to use any value you like for testing this issue:

https://github.com/FNA-XNA/FNA3D/blob/master/src/FNA3D_Driver_Vulkan.c#L5963

Even though we don't expect timeouts the renderer is prepared to handle VK_TIMEOUT as an acquireResult. FNA3D only depends on SDL, so provided the system has an SDL2-devel package installed it should build and run on anything as new as CentOS 7.

**Comment by fooishbar** [comment]
Which Vulkan driver are you using, and which version of it? Mesa has a bunch of code to only wait in ANI for as long as the timeout and return `VK_NOT_READY` (as distinct from `VK_TIMEOUT`) if no images are available within the given time.

**Comment by flibitijibibo** [comment]
> Which Vulkan driver are you using, and which version of it? Mesa has a bunch of code to only wait in ANI for as long as the timeout and return `VK_NOT_READY` (as distinct from `VK_TIMEOUT`) if no images are available within the given time.

This was tested on RADV, AMDVLK, and lavapipe.

**Comment by icculus** [comment]
I'm putting this in the 2.0.16 milestone, because it tops off a ton of really good Wayland work going into this release. I'm going to poke at the EGL code.

**Comment by icculus** [comment]
So do we have any opinions on framerate in eglSwapBuffers? We'll wait on the compositor, and if it's not coming through at all because it's ignoring an occluded window, we give up after 100 milliseconds (10fps), so the app can continue to tick?

Too low?

...too high?

**Comment by flibitijibibo** [comment]
My vote is for 10 frames' worth of time, based on the display's refresh_rate value.

**Comment by icculus** [comment]
Here's a first shot at this, and while this definitely gets the thing unblocked, with a swap interval of 1, testgl2 (a spinning cube) gets ~30fps here, like it's missing the first vsync. It's not clear to me if this is because I'm doing something dumb, it's just my laptop, or if this is the latency I was worried about.

Basically we just force the real eglSwapInterval to zero, listen for wl_surface_frame callbacks, set a flag when one comes in. When swapping buffers, if the app wants a swap interval, we wait for that flag to get set or timeout after 100 milliseconds (~10fps). We could probably put a sleep in the waiting code, or poll the Wayland display socket with a timeout to save CPU energy, but we don't at the moment.

Hitting 'h' in testgl2 hides the window, where it still makes progress every ~100ms (good), but just running it without touching anything makes it hit about 30fps (bad). It hits 30fps even if I comment out the debug printf calls, so it isn't that. Hitting 'o' will decrement the swap interval (which starts at 1), hitting 'p' will make it increment.

Basically everything not zero is adaptive vsync, since we either are waiting for the callback or the callback already happened.

```diff
diff --git a/include/SDL_video.h b/include/SDL_video.h
index 65fd6437..41c3e10b 100644
--- a/include/SDL_video.h
+++ b/include/SDL_video.h
@@ -1795,7 +1795,7 @@ extern DECLSPEC void SDLCALL SDL_GL_GetDrawableSize(SDL_Window * window, int *w,
  * vsync. Adaptive vsync works the same as vsync, but if you've already missed
  * the vertical retrace for a given frame, it swaps buffers immediately, which
  * might be less jarring for the user during occasional framerate drops. If
- * application requests adaptive vsync and the system does not support it,
+ * an application requests adaptive vsync and the system does not support it,
  * this function will fail and return -1. In such a case, you should probably
  * retry the call with 1 for the interval.
  *
diff --git a/src/video/wayland/SDL_waylandopengles.c b/src/video/wayland/SDL_waylandopengles.c
index 5da365c8..96be68c1 100644
--- a/src/video/wayland/SDL_waylandopengles.c
+++ b/src/video/wayland/SDL_waylandopengles.c
@@ -22,6 +22,8 @@
 
 #if SDL_VIDEO_DRIVER_WAYLAND && SDL_VIDEO_OPENGL_EGL
 
+#include "SDL_timer.h"
+#include "../../core/unix/SDL_poll.h"
 #include "../SDL_sysvideo.h"
 #include "../../events/SDL_windowevents_c.h"
 #include "SDL_waylandvideo.h"
@@ -55,19 +57,86 @@ Wayland_GLES_CreateContext(_THIS, SDL_Window * window)
     SDL_GLContext context;
     context = SDL_EGL_CreateContext(_this, ((SDL_WindowData *) window->driverdata)->egl_surface);
     WAYLAND_wl_display_flush( ((SDL_VideoData*)_this->driverdata)->display );
-    
     return context;
 }
 
+/* Wayland wants to tell you when to provide new frames, and if you have a non-zero
+   swap interval, Mesa will block until a callback tells it to do so. On some
+   compositors, they might decide that a minimized window _never_ gets a callback,
+   which causes apps to hang during swapping forever. So we always set the official
+   eglSwapInterval to zero to avoid blocking inside EGL, and manage this ourselves.
+   If a swap blocks for too long waiting on a callback, we just go on, under the
+   assumption the frame will be wasted, but this is better than freezing the app.
+   I frown upon platforms that dictate this sort of control inversion (the callback
+   is intended for _rendering_, not stalling until vsync), but we can work around
+   this for now.  --ryan. */
+/* Addendum: several recent APIs demand this sort of control inversion: Emscripten,
+   libretro, Wayland, probably others...it feels like we're eventually going to have
+   to give in with a future SDL API revision, since we can bend the other APIs to
+   this style, but this style is much harder to bend the other way.  :/ */
+int
+Wayland_GLES_SetSwapInterval(_THIS, int interval)
+{
+    if (!_this->egl_data) {
+        return SDL_SetError("EGL not initialized");
+    }
+    
+    /* technically, this is _all_ adaptive vsync (-1), because we can't
+       actually wait for the _next_ vsync if you set 1, but things that
+       request 1 probably won't care _that_ much. I hope. No matter what
+       you do, though, you never see tearing on Wayland. */
+    if (interval > 1) {
+        interval = 1;
+    } else if (interval < -1) {
+        interval = -1;
+    }
+
+    /* !!! FIXME: technically, this should be per-context, right? */
+    _this->egl_data->egl_swapinterval = interval;
+    _this->egl_data->eglSwapInterval(_this->egl_data->egl_display, 0);
+    return 0;
+}
+
+int
+Wayland_GLES_GetSwapInterval(_THIS)
+{
+    if (!_this->egl_data) {
+        SDL_SetError("EGL not initialized");
+        return 0;
+    }
+
+    return _this->egl_data->egl_swapinterval;
+}
+
 int
 Wayland_GLES_SwapWindow(_THIS, SDL_Window *window)
 {
     SDL_WindowData *data = (SDL_WindowData *) window->driverdata;
+    const int swap_interval = _this->egl_data->egl_swapinterval;
+
+    /* Control swap interval ourselves. See comments on Wayland_GLES_SetSwapInterval */
+    if (swap_interval != 0) {
+        const Uint32 max_wait = SDL_GetTicks() + 100;  /* ~10 fps, so we'll progress even if throttled to zero. */
+        struct wl_display *display = ((SDL_VideoData *)_this->driverdata)->display;
+        while ((SDL_AtomicGet(&data->swap_interval_ready) == 0) && (!SDL_TICKS_PASSED(SDL_GetTicks(), max_wait))) {
+            /* !!! FIXME: this is just the crucial piece of Wayland_PumpEvents */
+            WAYLAND_wl_display_flush(display);
+            if (SDL_IOReady(WAYLAND_wl_display_get_fd(display), SDL_FALSE, 0)) {
+                WAYLAND_wl_display_dispatch(display);
+            } else {
+                WAYLAND_wl_display_dispatch_pending(display);
+            }
+        }
+        SDL_AtomicSet(&data->swap_interval_ready, 0);
+    }
 
-    if (SDL_EGL_SwapBuffers(_this, data->egl_surface) < 0) {
-        return -1;
+    /* Feed the frame to Wayland. This will set it so the wl_surface_frame callback can fire again. */
+    if (!_this->egl_data->eglSwapBuffers(_this->egl_data->egl_display, data->egl_surface)) {
+        return SDL_EGL_SetError("unable to show color buffer in an OS-native window", "eglSwapBuffers");
     }
 
+printf("SWAP BUFFERS (INTERVAL %d) DONE %u\n", swap_interval, (unsigned int) SDL_GetTicks()); fflush(stdout);
+
     // Wayland-EGL forbids drawing calls in-between SwapBuffers and wl_egl_window_resize
     Wayland_HandlePendingResize(window);
 
@@ -89,7 +158,9 @@ Wayland_GLES_MakeCurrent(_THIS, SDL_Window * window, SDL_GLContext context)
     }
     
     WAYLAND_wl_display_flush( ((SDL_VideoData*)_this->driverdata)->display );
-    
+
+    _this->egl_data->eglSwapInterval(_this->egl_data->egl_display, 0);  /* see comments on Wayland_GLES_SetSwapInterval. */
+
     return ret;
 }
 
diff --git a/src/video/wayland/SDL_waylandopengles.h b/src/video/wayland/SDL_waylandopengles.h
index cfa72e61..0dfe512d 100644
--- a/src/video/wayland/SDL_waylandopengles.h
+++ b/src/video/wayland/SDL_waylandopengles.h
@@ -35,11 +35,11 @@ typedef struct SDL_PrivateGLESData
 #define Wayland_GLES_GetAttribute SDL_EGL_GetAttribute
 #define Wayland_GLES_GetProcAddress SDL_EGL_GetProcAddress
 #define Wayland_GLES_UnloadLibrary SDL_EGL_UnloadLibrary
-#define Wayland_GLES_SetSwapInterval SDL_EGL_SetSwapInterval
-#define Wayland_GLES_GetSwapInterval SDL_EGL_GetSwapInterval
 
 extern int Wayland_GLES_LoadLibrary(_THIS, const char *path);
 extern SDL_GLContext Wayland_GLES_CreateContext(_THIS, SDL_Window * window);
+extern int Wayland_GLES_SetSwapInterval(_THIS, int interval);
+extern int Wayland_GLES_GetSwapInterval(_THIS);
 extern int Wayland_GLES_SwapWindow(_THIS, SDL_Window * window);
 extern int Wayland_GLES_MakeCurrent(_THIS, SDL_Window * window, SDL_GLContext context);
 extern void Wayland_GLES_GetDrawableSize(_THIS, SDL_Window * window, int * w, int * h);
diff --git a/src/video/wayland/SDL_waylandwindow.c b/src/video/wayland/SDL_waylandwindow.c
index c06616bd..28c37849 100644
--- a/src/video/wayland/SDL_waylandwindow.c
+++ b/src/video/wayland/SDL_waylandwindow.c
@@ -22,7 +22,7 @@
 #include "../../SDL_internal.h"
 
 #if SDL_VIDEO_DRIVER_WAYLAND && SDL_VIDEO_OPENGL_EGL
-
+#include "SDL_timer.h"
 #include "../SDL_sysvideo.h"
 #include "../../events/SDL_windowevents_c.h"
 #include "../SDL_egl_c.h"
@@ -199,6 +199,24 @@ static const struct wl_shell_surface_listener shell_surface_listener_wl = {
 };
 
 
+static const struct wl_callback_listener surface_frame_listener;
+
+static void
+handle_surface_frame_done(void *data, struct wl_callback *cb, uint32_t time)
+{
+    SDL_WindowData *wind = (SDL_WindowData *) data;
+    SDL_AtomicSet(&wind->swap_interval_ready, 1);  /* mark window as ready to present again. */
+
+//printf("SURFACE FRAME CALLBACK! %u\n", (unsigned int) SDL_GetTicks()); fflush(stdout);
+
+    /* reset this callback to fire again once a new frame was presented and compositor wants the next one. */
+    wl_callback_destroy(cb);
+    wl_callback_add_listener(wl_surface_frame(wind->surface), &surface_frame_listener, data);
+}
+
+static const struct wl_callback_listener surface_frame_listener = {
+    handle_surface_frame_done
+};
 
 
 static void
@@ -996,6 +1014,9 @@ int Wayland_CreateWindow(_THIS, SDL_Window *window)
         wl_compositor_create_surface(c->compositor);
     wl_surface_add_listener(data->surface, &surface_listener, data);
 
+    /* fire a callback when the compositor wants a new frame rendered. */
+    wl_callback_add_listener(wl_surface_frame(data->surface), &surface_frame_listener, data);
+
 #ifdef SDL_VIDEO_DRIVER_WAYLAND_QT_TOUCH
     if (c->surface_extension) {
         data->extended_surface = qt_surface_extension_get_extended_surface(
diff --git a/src/video/wayland/SDL_waylandwindow.h b/src/video/wayland/SDL_waylandwindow.h
index baddeabb..2312df27 100644
--- a/src/video/wayland/SDL_waylandwindow.h
+++ b/src/video/wayland/SDL_waylandwindow.h
@@ -59,6 +59,7 @@ typedef struct {
         SDL_zxdg_shell_surface zxdg;
         struct wl_shell_surface *wl;
     } shell_surface;
+
     struct wl_egl_window *egl_window;
     struct SDL_WaylandInput *keyboard_device;
     EGLSurface egl_surface;
@@ -67,6 +68,8 @@ typedef struct {
     struct zwp_keyboard_shortcuts_inhibitor_v1 *key_inhibitor;
     struct zwp_idle_inhibitor_v1 *idle_inhibitor;
 
+    SDL_atomic_t swap_interval_ready;
+
 #ifdef SDL_VIDEO_DRIVER_WAYLAND_QT_TOUCH
     struct qt_extended_surface *extended_surface;
 #endif /* SDL_VIDEO_DRIVER_WAYLAND_QT_TOUCH */
diff --git a/test/testgl2.c b/test/testgl2.c
index 50561049..3d892a60 100644
--- a/test/testgl2.c
+++ b/test/testgl2.c
@@ -301,6 +301,9 @@ main(int argc, char *argv[])
         swap_interval = 0;
     }
 
+SDL_GL_SetSwapInterval(1);
+swap_interval = 1;
+
     SDL_GetCurrentDisplayMode(0, &mode);
     SDL_Log("Screen BPP    : %d\n", SDL_BITSPERPIXEL(mode.format));
     SDL_Log("Swap Interval : %d\n", SDL_GL_GetSwapInterval());
@@ -389,12 +392,16 @@ main(int argc, char *argv[])
         while (SDL_PollEvent(&event)) {
             SDLTest_CommonEvent(state, &event, &done);
             if (event.type == SDL_KEYDOWN) {
-                if (event.key.keysym.sym == SDLK_o) {
+                if (event.key.keysym.sym == SDLK_h) {
+                    SDL_MinimizeWindow(state->windows[0]);
+                } else if (event.key.keysym.sym == SDLK_o) {
                     swap_interval--;
                     update_swap_interval = SDL_TRUE;
                 } else if (event.key.keysym.sym == SDLK_p) {
                     swap_interval++;
                     update_swap_interval = SDL_TRUE;
+                } else if (event.key.keysym.sym == SDLK_h) {
+                    SDL_HideWindow(state->windows[0]);
                 }
             }
         }
@@ -414,7 +421,9 @@ main(int argc, char *argv[])
             SDL_GL_GetDrawableSize(state->windows[i], &w, &h);
             ctx.glViewport(0, 0, w, h);
             Render();
+//printf("SWAP WINDOW!\n"); fflush(stdout);
             SDL_GL_SwapWindow(state->windows[i]);
+//printf("SWAP WINDOW DONE!\n"); fflush(stdout);
         }
     }
```

**Comment by icculus** [comment]
Anyhow, that's kinda a verbose patch to ask someone to sort through and pick out what I did wrong, but I don't know why this is only able to hit 30fps, and would appreciate any insight.

**Comment by icculus** [comment]
Small update: gnome-shell on my laptop is taking 80% of the CPU while the system is apparently idle, so I'm guessing this is the culprit.

Someone with a more trustworthy Wayland setup should try my patch, I guess.

**Comment by christianrauch** [comment]
> Here's a first shot at this, and while this definitely gets the thing unblocked, with a swap interval of 1, testgl2 (a spinning cube) gets ~30fps here

The examples I looked at set the swap interval to 0. I think this is the expected usage to not have the swap to block. The render loop should be entirely driven by the frame callback without the need to delay anything manually. 

As far as I understand OpenGL, an interval of 1 will always wait for one additional frame, which halves the frames shown per second (e.g. you skip every other frame). On a 120Hz screen, you will probably see a framerate of 60Hz this way. A swap interval of 2 would probably get you 20Hz of framerate in the example.

>  gnome-shell on my laptop is taking 80%

This is odd. Does it take that much when running the OpenGL example? Or already in idle mode, e.g. with a single window?
Mine is using ~5% idle (of one CPU), so 80% is not normal. It should only ever rise that much when doing a lot of window managing and compositing work, such as when toggling maximised/floating at multiple times a second. But even then, on my system, this is only taking 50% of a CPU.

Can you provide a PR? This makes it easier for me to check it out locally and provide a PR ontop of it to fix any issues.

**Comment by flibitijibibo** [comment]
>As far as I understand OpenGL, an interval of 1 will always wait for one additional frame, which halves the frames shown per second (e.g. you skip every other frame). On a 120Hz screen, you will probably see a framerate of 60Hz this way. A swap interval of 2 would probably get you 20Hz of framerate in the example.

This is not how it works - the math is `refresh rate / swap interval` when interval is > 0. 0 burns without waiting, and -1 is a special value that acts like 1, except if you go over the refresh rate to render it skips waiting so you don't get the same behavior as interval 2.

**Comment by icculus** [comment]
> This is odd. Does it take that much when running the OpenGL example? Or already in idle mode, e.g. with a single window?
Mine is using ~5% idle (of one CPU), so 80% is not normal.

Yeah, there's something wrong with my setup, unrelated to OpenGL, Wayland, or SDL. It's doing it on a fresh boot with no applications launched, too. It's about time to blow away this install and start over.   :)

**Comment by icculus** [comment]
> and -1 is a special value that acts like 1, except if you go over the refresh rate to render it skips waiting so you don't get the same behavior as interval 2.

...and in Wayland's case, everything that's not zero acts like -1, since we don't know when the actual retrace is, we can only wait for the compositor to request a new frame (we're at vsync) or feed it a frame immediately because it already asked for one (we missed the vsync).

Zero, of course, just dumps the frame on the compositor with reckless abandon, even if its wasted.

**Comment by fooishbar** [comment]
> ...and in Wayland's case, everything that's not zero acts like -1, since we don't know when the actual retrace is, we can only wait for the compositor to request a new frame

FWIW we do have things in the works for this, but it’s tied in with the still-WIP Vulkan extension since we want to just do it right once.

(For extreme nitpick, ‘retrace’ isn’t relevant: what’s relevant is when the compositor decides which buffer to latch for a given frame, which may be significantly earlier than actual retrace if it needs to schedule GPU work for composition. And by ‘frame’ I mean ‘present cycle’ because hello VRR ... anyway.)

**Comment by flibitijibibo** [comment]
Looking at the patch, the 30fps bit makes sense I think - if it's going to wait for sync it will probably do this for both windows, so each window will alternate getting vblank, meaning they run at 30 in alternating order. This would probably be an issue with other GL WSI as well.

**Comment by icculus** [comment]
> for both windows

To be clear: I was running with one window.

**Comment by flibitijibibo** [comment]
Oh, woops... yeah, that is odd. At the moment my office is covered up so I can't do a local test, probably will be stuck until Monday.

**Comment by icculus** [comment]
Update: tried it on KDE on same machine, it hits 60fps, so this is a Gnome bug (that is likely either something broken on my machine or something they've fixed since I installed Ubuntu 19.04 and never upgraded).

I guess I should test that multiple windows don't get stuck, since that's the actual bug in question, but then I think it's safe to push this change.

**Comment by flibitijibibo** [comment]
I only see two possible issues:

- We probably need to guard this for SDL_WINDOW_OPENGL, Vulkan is basically screwed for now
- Do we want to use atomics for this? My main worry is that games with really low frame times will end up burning the CPU while it waits for the frame callback to fire.

Aside from that, I guess I can't complain if it works around this. Not looking forward to figuring out why Vulkan timeouts don't work here.

**Comment by fooishbar** [comment]
> We probably need to guard this for SDL_WINDOW_OPENGL, Vulkan is basically screwed for now

Vulkan WSI doesn't synchronise against `frame` if you use mailbox rather than FIFO.

**Comment by flibitijibibo** [comment]
> > We probably need to guard this for SDL_WINDOW_OPENGL, Vulkan is basically screwed for now
> 
> Vulkan WSI doesn't synchronise against `frame` if you use mailbox rather than FIFO.

True, but that brings us back to "screw it, let's burn the CPU," which to be fair I'm up for doing if it means I don't have to think about two schedulers instead of just one.

**Comment by icculus** [comment]
Pushed my attempt at this. As noted, if you have two windows, the main window is going to run at 10 fps because the hidden one is blocking for 100ms before the other gets a chance to do so (unless they swapbuffers from different threads).

We can improve upon this situation in various ways, but I think this is Good Enough for now, as the app will no longer fail to make progress at all.

**Comment by flibitijibibo** [comment]
To take a swing at this dead horse one more time, I've updated wltest with the latest SDL Wayland work and and equivalent timeout for FNA3D Vulkan:

https://github.com/FNA-XNA/FNA3D/commit/083717cdf985445720bb4867908195daf8c816e2

http://flibitijibibo.com/wltest.tar.bz2

Still doesn't work in Plasma after hitting the space bar, GNOME is unaffected as it always worked before the timeout patchsets anyway.
